stratosphere/pm/source/impl/pm_process_info.hpp

```diff
@@ -0,0 +1,310 @@
+            }
+
+            bool HasExited() const {
+                return this->state == ProcessState_Exited;
+            }
+
+            bool ShouldSignalOnExit() const {
+                return this->HasFlag(Flag_SignalOnExit);
+            }
+
+            bool HasExceptionOccurred() const {
+                return this->HasFlag(Flag_ExceptionOccurred);
+            }
+
+            bool HasExceptionWaitingAttach() const {
+                return this->HasFlag(Flag_ExceptionWaitingAttach);
+            }
+
+            bool ShouldSignalOnDebugEvent() const {
+                return this->HasFlag(Flag_SignalOnDebugEvent);
+            }
+
+            bool ShouldSignalOnStart() const {
+                return this->HasFlag(Flag_SignalOnStart);
+            }
+
+            bool HasSuspendedStateChanged() const {
+                return this->HasFlag(Flag_SuspendedStateChanged);
+            }
+
+            bool IsSuspended() const {
+                return this->HasFlag(Flag_Suspended);
+            }
+
+            bool IsApplication() const {
+                return this->HasFlag(Flag_Application);
+            }
+
+            bool HasStartedStateChanged() const {
+                return this->HasFlag(Flag_StartedStateChanged);
+            }
+
+    };
+
+    Result OnProcessSignaled(std::shared_ptr<ProcessInfo> process_info);
+
+    class ProcessInfoWaiter final : public IWaitable {
+        private:
+            std::shared_ptr<ProcessInfo> process_info;
+        public:
+            ProcessInfoWaiter(std::shared_ptr<ProcessInfo> p) : process_info(p) { /* ... */ }
+
+            /* IWaitable */
+            Handle GetHandle() override {
+                return this->process_info->GetHandle();
+            }
+
+            Result HandleSignaled(u64 timeout) override {
+                return OnProcessSignaled(this->process_info);
+            }
+    };
+
+    class ProcessList final {
+        private:
+            HosMutex lock;
+            std::vector<std::shared_ptr<ProcessInfo>> processes;
+        public:
+            void Lock() {
+                this->lock.Lock();
+            }
+
+            void Unlock() {
+                this->lock.Unlock();
+            }
+
+            size_t GetSize() const {
+                return this->processes.size();
+            }
+
+            std::shared_ptr<ProcessInfo> Pop() {
+                auto front = this->processes[0];
+                this->processes.erase(this->processes.begin());
+                return front;
+            }
+
+            void Add(std::shared_ptr<ProcessInfo> process_info) {
+                this->processes.push_back(process_info);
+            }
+
+            void Remove(u64 process_id) {
+                for (size_t i = 0; i < this->processes.size(); i++) {
+                    if (this->processes[i]->GetProcessId() == process_id) {
+                        this->processes.erase(this->processes.begin() + i);
```

Comment:
use an iterator to avoid `this->processes.begin() + i`?

Reply:
ah, yeah, good call.